### PR TITLE
V1r2

### PR DESCRIPTION
**Pull request by wdower**
Initializing profile repo at v1r2.

### CODE REVIEW

**Review by Amndeep7 — CHANGES_REQUESTED** [review]

File: controls/SV-257811.rb [hunk 1]
```diff
@@ -26,15 +26,60 @@
...
+
+    # Check the configuration files
+    describe 'Configuration files' do
+      if search_results.empty?
+        it "do not explicitly set the `#{parameter}` parameter" do
+          expect(search_results).not_to be_empty, "Add the line `#{parameter}=#{value}` to a file in the `/etc/sysctl.d/` directory"
+        end
+      else
+        it "set the parameter to the right value for #{action}" do
+          expect(search_results).to match(/#{parameter}\s*=\s*#{value}/)
```
Comment (Amndeep7) [inline controls/SV-257811.rb:80]: These regexes have also been highly inconsistent.  Here we are only looking for the parameter name then any whitespace then equal then any whitespace then the value.  In other places we don't allow for whitespace or we ensure that the value is at the end of the line.  IMO it should be closer to all of the above lol: `/^\s*#{parameter}\s*=\s*#{value}$/`

File: .github/workflows/verify-container.yml
Comment (Amndeep7) [inline .github/workflows/verify-container.yml:1]: image

we should resolve all those warnings probably.  at the minimum at least bump the versions of all the actions we're using.

this applies to all the workflows.

**Review by Amndeep7 — CHANGES_REQUESTED** [review]

File: controls/SV-257826.rb [hunk 2]
```diff
@@ -15,15 +13,24 @@
...
-  tag satisfies: ['SRG-OS-000074-GPOS-00042', 'SRG-OS-000095-GPOS-00049', 'SRG-OS-000480-GPOS-00227']
-  tag 'documentable'
-  tag cci: ['CCI-000197', 'CCI-000366', 'CCI-000381']
-  tag nist: ['IA-5 (1) (c)', 'CM-6 b', 'CM-7 a']
+  tag cci: ['CCI-000366', 'CCI-000197', 'CCI-000381']
+  tag nist: ['CM-6 b', 'IA-5 (1) (c)', 'CM-7 a']
+  tag 'host'
+  tag 'container'
+
+  if input('ftp_required')
```
Comment (Amndeep7) [inline controls/SV-257826.rb:27]: why is there an exception for ftp when there is no mention of an isso waiver?  have we had an ask for this in the past?

File: controls/SV-257840.rb [hunk 3]
```diff
@@ -25,4 +25,9 @@
   tag 'documentable'
   tag cci: ['CCI-000366']
   tag nist: ['CM-6 b']
+  tag 'host', 'container'
```
Comment (Amndeep7) [inline controls/SV-257840.rb:28]: nitpick: should be on separate lines
```suggestion
  tag 'host'
  tag 'container'
```

File: controls/SV-257842.rb [hunk 4]
```diff
@@ -17,14 +17,22 @@
...
   tag stig_id: 'RHEL-09-215095'
   tag gtitle: 'SRG-OS-000363-GPOS-00150'
   tag fix_id: 'F-61507r925512_fix'
   tag 'documentable'
   tag cci: ['CCI-001744']
   tag nist: ['CM-3 (5)']
+  tag 'host'
+  tag 'container'
+
+  mail_package = input('mail_package')
```
Comment (Amndeep7) [inline controls/SV-257842.rb:33]: the value for this input in the inspec.yml is 'mailx' but the control says 's-nail'.  we should either better document the reason for our override in the default values (and probably have 's-nail' there but commented out) or correct the default value to be 's-nail'.

File: controls/SV-257843.rb [hunk 5]
```diff
@@ -11,14 +11,45 @@
...
   tag rid: 'SV-257843r925516_rule'
   tag stig_id: 'RHEL-09-231010'
-  tag gtitle: 'SRG-OS-000480-GPOS-00227'
   tag fix_id: 'F-61508r925515_fix'
-  tag 'documentable'
   tag cci: ['CCI-000366']
   tag nist: ['CM-6 b']
+  tag 'host'
+
+  only_if('This requirement is Not Applicable inside a container, the containers host manages the containers filesystems') {
```
Comment (Amndeep7) [inline controls/SV-257843.rb:24]:
```suggestion
  only_if('This requirement is Not Applicable inside a container, the container's host manages the container's filesystems') {
```

File: controls/SV-257844.rb [hunk 6]
```diff
@@ -11,14 +11,25 @@
...
+  only_if('This control is Not Applicable to containers', impact: 0.0) {
+    !virtualization.system.eql?('docker')
+  }
+
+  describe mount('/tmp') do
+    it { should be_mounted }
+  end
+
+  describe etc_fstab.where { mount_point == '/tmp' } do
+    it { should exist }
```
Comment (Amndeep7) [inline controls/SV-257844.rb:33]: https://docs.chef.io/inspec/resources/etc_fstab/#check-if-a-partition-is-mounted-at-a-point

seemed like it ought to be be_configured not just exist

this comment applies to all the controls involving checking if a directory has its own partition, ex. so the home dir one too

File: controls/SV-257847.rb [hunk 7]
```diff
@@ -15,15 +13,27 @@
...
-  tag 'documentable'
   tag cci: ['CCI-000366', 'CCI-001849']
   tag nist: ['CM-6 b', 'AU-4']
+  tag 'host'
+
+  only_if('This control is Not Applicable to containers', impact: 0.0) {
+    !virtualization.system.eql?('docker')
+  }
+
+  audit_data_path = command("dirname #{auditd_conf.log_file}").stdout.strip
```
Comment (Amndeep7) [inline controls/SV-257847.rb:30]: i cannot find this auditd_conf input in the inspec.yml nor is it used in any of the sample/testing inputs.ymls.

File: controls/SV-257848.rb
Comment (Amndeep7) [inline controls/SV-257848.rb:1]: do we not need to check if it is mounted like in the other controls?

File: controls/SV-257850.rb [hunk 8]
```diff
@@ -25,4 +25,37 @@
...
   tag cci: ['CCI-001764']
   tag nist: ['CM-7 (2)']
+  tag 'host'
+
+  only_if('Control not applicable within a container', impact: 0.0) {
+    !virtualization.system.eql?('docker')
+  }
+
+  interactive_users = passwd.where {
+    uid.to_i >= 1000 && shell !~ /nologin/
```
Comment (Amndeep7) [inline controls/SV-257850.rb:35]: here we are just checking that our shell does not match nologin, but SV-257843 checks against a list of non-interactive shells.  do we not need to check against that full list here?

**Review by Amndeep7 — CHANGES_REQUESTED** [review]

File: .github/workflows/verify-container.yml [hunk 9]
```diff
@@ -94,7 +91,7 @@ jobs:
         if: ${{ !contains(steps.commit.outputs.message, 'only-validate-profile') }}
         continue-on-error: true
         run: |
-          curl -# -s -F data=@spec/results/${{ env.PLATFORM }}_${{ matrix.suite }}.json -F "filename=${{ env.COMMIT_SHORT_SHA }}-${{ env.PLATFORM }}_${{ matrix.suite }}" -F "public=true" -F "evaluationTags=${{ env.COMMIT_SHORT_SHA }},${{ github.repository }},${{ github.workflow }}" -H "Authorization: Api-Key ${{ secrets.HEIMDALL_UPLOAD_GROUP_KEY }}" "${{ vars.SAF_HEIMDALL_URL }}/evaluations"
+          curl -# -s -F data=@spec/results/${{ env.PLATFORM }}_${{ matrix.suite }}.json -F "filename=${{ env.COMMIT_SHORT_SHA }}-${{ env.PLATFORM }}_${{ matrix.suite }}" -F "public=true" -F "evaluationTags=${{ env.COMMIT_SHORT_SHA }},${{ github.repository }},${{ github.workflow }}" -H "Authorization: Api-Key ${{ secrets.SAF_HEIMDALL_UPLOAD_GROUP_KEY }}" "${{ vars.SAF_HEIMDALL_URL }}/evaluations"
```
[the remark this answers is not in the document]
Comment (wdower) [reply .github/workflows/verify-container.yml:94]: SAF_HEIMDALL_URL is a variable and not a secret. It's set at the org level.

[the remark this answers is not in the document]
Comment (wdower) [reply .github/workflows/verify-container.yml:94]: Updating the var name.

File: README.md [hunk 10]
```diff
@@ -1,9 +1,9 @@
 # (WIP) RedHat Enterprise Linux 9.x Security Technical Implementation Guide InSpec Profile
```
[the remark this answers is not in the document]
Comment (wdower) [reply README.md:1]: Good catch.

File: .github/workflows/lint-profile.yml [hunk 11]
```diff
@@ -1,9 +1,6 @@
 name: Lint & Check the Profile
 
 on:
-  push:
```
[the remark this answers is not in the document]
Comment (wdower) [reply .github/workflows/lint-profile.yml:4]: Fair.

File: .github/workflows/verify-container.yml
Comment (wdower) [reply .github/workflows/verify-container.yml:1]: Not clear on what the point of that pipeline step even was. Striking it.

File: controls/SV-257777.rb [hunk 12]
```diff
@@ -13,14 +13,43 @@
...
+  EOMS_DATE = {
+    /^9\.0/ => '31 May 2024',
+    /^9\.1/ => 'April 1, 2023',
+    /^9\.2/ => 'May 31, 2025',
+    /^9\.4/ => 'May 31, 2026'
+  }.find { |k, _v| k.match(release) }&.last
+
+  describe "The release \"#{release}\"" do
+    if EOMS_DATE.nil?
+      it 'is a supported release' do
```
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257777.rb:46]: No. `The release is a supported release` is what the test will print in the output, and the test will fail because EOMS_DATE is a nil (ergo its not a release with a support window published). We're forcing a fail if EOMS_DATE is nil but doing so such that the user will understand what the problem was.

File: controls/SV-257779.rb [hunk 13]
```diff
@@ -47,15 +45,36 @@
...
+  end
+
+  if banner_file.exist?
+
+    banner = banner_file.content.gsub(/[\r\n\s]/, '')
+    expected_banner = input('banner_message_text_cli').gsub(/[\r\n\s]/, '')
+
+    describe 'The CLI Login Banner ' do
+      it 'is set to the standard banner and has the correct text' do
+        expect(banner).to eq(expected_banner), 'Banner does not match expected text'
```
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257779.rb:76]: The whitespace thing was the easiest way of determining if the banner messages match in content; we do it like this for all the linux OSes. It's clunky but there's not too many ways of doing better string comparison in a way that makes sense in the output.

File: controls/SV-257782.rb [hunk 14]
```diff
@@ -19,14 +21,29 @@
...
   tag nist: ['CM-6 b']
+  tag 'host'
+
+  only_if('This control is Not Applicable to containers', impact: 0.0) {
+    !virtualization.system.eql?('docker')
+  }
+
+  if os.release.to_f >= 8.4 && input('use_fips') == true
+    impact 0.0
+    describe 'For RHEL versions 8.4 and above running with kernel FIPS mode enabled as specified by RHEL-08-010020, this requirement is Not Applicable.' do
```
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257782.rb:40]: Good catch.

File: controls/SV-257788.rb [hunk 15]
```diff
@@ -23,4 +23,15 @@
   tag 'documentable'
   tag cci: ['CCI-000366']
   tag nist: ['CM-6 b']
+  tag 'host'
+
+  only_if('Control not applicable within a container without sudo enabled', impact: 0.0) do
+    !virtualization.system.eql?('docker')
+  end
+
+  grubby = command('grubby --info=ALL').stdout
```
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257788.rb:32]: It's a default.

File: controls/SV-257792.rb [hunk 16]
```diff
@@ -38,4 +38,19 @@
   tag 'documentable'
   tag cci: ['CCI-000366', 'CCI-001084']
   tag nist: ['CM-6 b', 'SC-3']
+  tag 'host'
+
+  only_if('Control not applicable within a container', impact: 0.0) {
+    !virtualization.system.eql?('docker')
+  }
+
+  grub_stdout = command('grub2-editenv - list').stdout
```
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257792.rb:47]: Yup, looks like this logic should have changed between 8 and 9. Have updated it.

File: controls/SV-257797.rb [hunk 17]
```diff
@@ -37,15 +35,39 @@
...
+    !virtualization.system.eql?('docker')
+  }
+
+  action = 'kernel.dmesg_restrict'
+
+  describe kernel_parameter(action) do
+    its('value') { should eq 1 }
+  end
+
+  search_result = command("grep -r ^#{action} #{input('sysctl_conf_files').join(' ')}").stdout.strip
```
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257797.rb:58]: We can add wildcards to the input to search entire directories and I don't think the systemd conf files will change too much from there. The STIG check command also only lists the files as comments, which would make it very difficult to parse for the paths. I think we should stick with the pattern of "let the user tell me if we are deviating from reading default files."

File: controls/SV-257797.rb [hunk 18]
```diff
@@ -37,15 +35,39 @@
...
+
+  correct_result = search_result.lines.any? { |line| line.match(/#{action}\s*=\s*1$/) }
```
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257797.rb:60]: The grep already only matches on line start (ergo no comments). We could refactor but it wouldn't really be more efficient than it is now.

File: controls/SV-257800.rb
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257800.rb:1]: I prefer using the other pattern; I think it reads cleaner. This one was a holdover. Updated.

File: controls/SV-257800.rb [hunk 19]
```diff
@@ -25,15 +23,72 @@
...
+  # Define the kernel parameter to be checked
+  parameter = 'kernel.kptr_restrict'
+  action = 'kernel pointer addresses'
+  value = 1
+
+  # Get the current value of the kernel parameter
+  current_value = kernel_parameter(parameter)
+
+  # Check if the system is a Docker container
+  if virtualization.system.eql?('docker')
```
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257800.rb]: You can only use one `only_if` statement per control (had it confirmed as intentional by Chef). Hence we frequently do the only_if thing alongside the usual if/skip checks.

File: controls/SV-257803.rb
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257803.rb:1]: You're not wrong but I think updating the profile for consistency on that front would be beyond the scope of this PR.

Adding an issue: https://github.com/mitre/redhat-enterprise-linux-9-stig-baseline/issues/39

File: controls/SV-257803.rb [hunk 20]
```diff
@@ -27,14 +27,40 @@
...
+  }
+
+  kernel_setting = 'kernel.core_pattern'
+  kernel_expected_value = input('kernel_dump_expected_value')
+
+  describe kernel_parameter(kernel_setting) do
+    its('value') { should eq kernel_expected_value }
+  end
+
+  k_conf_files = input('kernel_config_files')
```
[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257803.rb]: Fixed.

[the remark this answers is not in the document]
Comment (wdower) [reply controls/SV-257803.rb]: Just checked and it seems that inspec's `kernel_module` gives the actively loaded kernel modules but not the contents of the conf files. These will need to have that check added: https://github.com/mitre/redhat-enterprise-linux-9-stig-baseline/issues/40

File: controls/SV-257889.rb [hunk 21]
```diff
@@ -19,14 +21,31 @@
...
+  tag 'host'
+
+  only_if('This control is Not Applicable to containers', impact: 0.0) {
+    !virtualization.system.eql?('docker')
+  }
+
+  ignore_shells = input('non_interactive_shells').join('|')
+
+  homedirs = users.where { !shell.match(ignore_shells) && (uid >= 1000 || uid.zero?) }.homes
+  ifiles = command("find #{homedirs.join(' ')} -xdev -maxdepth 1 -name '.*' -type f").stdout.split("\n")
```
Comment (Amndeep7) [inline controls/SV-257889.rb:41]: this seems like it's just trying to find all one's dotfiles in one's homedir.  nominally that's wrong because there could be dotfiles in there that don't impact the shell and are used for configuring other applications, ex. .vimrc for vim.

additionally, configuration files don't necessarily need to exist in just the top level.  for example, you could have your .bashrc at the top level, but then have it source further configuration files from another location that we wouldn't find with this `find`.

furthermore, there are even specifications such as freedesktop's [XDG Base Directory Specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) that have gotten increasing amounts of [support ](https://wiki.archlinux.org/title/XDG_Base_Directory )where people often set their configuration files - even for shells.

i think if we wanted to do this properly we would a) keep an eye on all the dotfiles in the homedir, b) have an input that would tell us of additional locations we would need to inspect, and c) have an input that would have a list of files that we could ignore

**Review by Amndeep7 — CHANGES_REQUESTED** [review]

File: controls/SV-257779.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (Amndeep7) [reply controls/SV-257779.rb:76]: @wdower yeah for the actual test that's fine.  I'm talking about doing a custom message here to change what shows up in the HDF file

File: controls/SV-257800.rb (hunk 19 above)
[the remark this answers is not in the document]
Comment (Amndeep7) [reply controls/SV-257800.rb]: Dang that's annoying.  Oh well.

File: controls/SV-257898.rb [hunk 22]
```diff
@@ -23,4 +23,10 @@
   tag 'documentable'
   tag cci: ['CCI-000366']
   tag nist: ['CM-6 b']
+  tag 'host', 'container'
```
Comment (Amndeep7) [inline controls/SV-257898.rb:26]:
```suggestion
  tag 'host'
  tag 'container'
```

**Review by Amndeep7 — CHANGES_REQUESTED** [review]

File: controls/SV-257928.rb [hunk 23]
```diff
@@ -17,15 +15,25 @@
...
-  tag satisfies: ['SRG-OS-000480-GPOS-00227', 'SRG-OS-000138-GPOS-00069']
-  tag 'documentable'
   tag cci: ['CCI-000366', 'CCI-001090']
   tag nist: ['CM-6 b', 'SC-4']
+  tag 'host'
+  tag 'container'
+
+  partitions = etc_fstab.params.map { |partition| partition['mount_point'] }.uniq
+
+  cmd = "find #{partitions.join(' ')} -xdev -type d -perm -0002 -uid +999 -print"
```
Comment (Amndeep7) [inline controls/SV-257928.rb:31]: Going back to my earlier comments asking if we actually need to call `-print` here:

> The -print action is performed on all files for which the whole expression is true, unless it contains an action other than -prune or -quit.  Actions which inhibit the default -print are -delete, -exec, -execdir, -ok, -okdir, -fls, -fprint, -fprintf, -ls, -print and -printf.

None of those are true here so we could probably get rid of the call to that parameter.  However, this is also in the man page: 

> If you are piping the output of find into another program and there is the faintest possibility that the files which you are searching for might contain a newline, then you should seriously consider using the -print0 option instead of -print.

Since this is explicitly what we're doing, should we swap all usages of print with print0?  The other code change we'd need to do is swap the character we're doing the split on from \n to \0 (the null character).

An argument against it is that in this specific instance anything that shows up is a failure, but what about the several controls when we're putting in exemptions from an input?  Is it possible that a filename could be constructed that could take advantage of this?

Doing some research, it seems like you can relatively easily append newlines to the end of a filename.  Splitting the results of -print on a newline would then allow for a malicious filename to get through.  

However, my non-exhaustive skim of the various finds we are doing didn't show anywhere that that could be taken advantage of unless the user's inputs file has been compromised in some manner or they've done a typo of some kind at which point we're boned regardless.

tl;dr you're probably better off not reading unless you like the musings of 3am brain.  it was at least fun to read some of the find manual.

**Review by Amndeep7 — CHANGES_REQUESTED** [review]
